### DISCUSSION TITLE
Problem sending large files as base64Binary from cxf client to cxf service

### DISCUSSION THREAD

**Discussion by xperjon** [quarkiverse/quarkus-cxf#971] [Q&A] [answered]
Hi!
I'm having problem sending large files from a cxf client to a cxf service. I made a a sample application to reproduce: https://github.com/xperjon/Quarkus-CXF-FileUpload

Small files works just fine but when I try larger ones (about 100kB) the server throws:

```
Caused by: io.netty.handler.codec.http2.Http2Exception: Flow control window exceeded for stream: 0
at io.netty.handler.codec.http2.Http2Exception.connectionError(Http2Exception.java:109)

```
and the client throws:
```
Caused by: jakarta.xml.bind.MarshalException
 - with linked exception:
[java.io.IOException: Read end dead]
at org.glassfish.jaxb.runtime.v2.runtime.MarshallerImpl.write(MarshallerImpl.java:268)
```

My wsdl looks like this (highly inspired by the  example at https://docs.quarkiverse.io/quarkus-cxf/dev/user-guide/index.html):
```
<wsdl:definitions name = "FileService"
   targetNamespace = "http://server.it.cxf.quarkiverse.io/"
   xmlns:wsdl = "http://schemas.xmlsoap.org/wsdl/"
   xmlns:soap = "http://schemas.xmlsoap.org/wsdl/soap/"
   xmlns:tns = "http://server.it.cxf.quarkiverse.io/"
   xmlns:xsd = "http://www.w3.org/2001/XMLSchema">
   
   <wsdl:types>
   <xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:tns="http://server.it.cxf.quarkiverse.io/" attributeFormDefault="unqualified" elementFormDefault="unqualified" targetNamespace="http://server.it.cxf.quarkiverse.io/">

   <xsd:element name="upload" type="tns:upload"/>   
   <xsd:complexType name="upload">
       <xsd:sequence>
           <xsd:element minOccurs="0" name="filename" type="xsd:string"/>
           <xsd:element minOccurs="0" name="data" type="xsd:base64Binary"/>
        </xsd:sequence>
    </xsd:complexType>

    <xsd:element name="uploadResponse" type="tns:uploadResponse"/>
   <xsd:complexType name="uploadResponse">
     <xsd:sequence>
       <xsd:element minOccurs="0" name="size" type="xsd:int"/>
     </xsd:sequence>
   </xsd:complexType>
 </xsd:schema>
</wsdl:types>

   <wsdl:message name = "upload">
      <wsdl:part element="tns:upload" name = "parameters"></wsdl:part>
   </wsdl:message >

   <wsdl:message name = "uploadResponse">
    <wsdl:part element="tns:uploadResponse" name = "parameters"></wsdl:part>
 </wsdl:message >


   <wsdl:portType name = "FileService">
      <wsdl:operation name = "uploadFile">
         <wsdl:input message = "tns:upload" name="upload"/>
         <wsdl:output message = "tns:uploadResponse" name="uploadResponse"/>
      </wsdl:operation>
   </wsdl:portType>

   <wsdl:binding name = "FileServiceSoapBinding" type = "tns:FileService">
      <soap:binding style = "document"
         transport = "http://schemas.xmlsoap.org/soap/http"/>
      <wsdl:operation name = "uploadFile">
         <soap:operation soapAction = "" style="document"/>
         <wsdl:input name="upload">
            <soap:body use="literal"/>
         </wsdl:input>
		
         <wsdl:output name="uploadResponse">
            <soap:body use="literal"/>
         </wsdl:output>
      </wsdl:operation>
   </wsdl:binding>

   <wsdl:service name = "FileService">
      <wsdl:documentation>WSDL File for FileService</wsdl:documentation>
      <wsdl:port binding = "tns:FileServiceSoapBinding" name = "File">
         <soap:address
            location = "http://localhost:8080/FileService" />
      </wsdl:port>
   </wsdl:service>
</wsdl:definitions>
```

I'm using Java17 and Quarkus CXF 3.2.4.Final on both client and service. 
Any ideas?

**Comment by shumonsharif** [comment]
Hi @xperjon It looks like HTTP/2 is getting used by the server, so the below may be one possible approach to resolving your issue.

Add the following dependency for client:

```xml
    <dependency>
      <groupId>io.quarkiverse.cxf</groupId>
      <artifactId>quarkus-cxf-rt-transports-http-hc5</artifactId>
    </dependency>
```

And then add the below to your client class:

```java
    @PostConstruct
    public void configureHttp2() {
        Bus bus = ClientProxy.getClient(fileService).getBus();
        bus.setProperty(AsyncHTTPConduit.USE_ASYNC, Boolean.TRUE);
        bus.setProperty(AsyncHTTPConduit.ENABLE_HTTP2, Boolean.TRUE);
    }
```

You can read more about the above here:  https://cxf.apache.org/docs/asynchronous-client-http-transport.html

**Reply by ppalaga** [reply]
Thanks for looking, @shumonsharif, but I doubt your suggestion can really help. I have not tried your suggestion, but I have run @xperjon 's reproducer and I can state the following:

* The exchange works well for smaller payloads, hence HTTP/2 should not be the problem
* For larger payloads, the exception is thrown from netty code on the server side. 
* Setting `quarkus.http.http2=false` on the server causes a slightly different exception on a different place in netty code, hence HTTP/1 does not work either.

This looks like a serious issue. Thanks for the report @xperjon and thanks a lot for the reproducer!

I am about to send a PR modifying our MTOM test to reproduce the issue and I will consult somebody from Quarkus team.

You can watch the progress in https://github.com/quarkiverse/quarkus-cxf/issues/973

**Reply by shumonsharif** [reply]
Thanks @ppalaga ... I'm not convinced either of us understand the root cause of the issue (yet) ... 

However, not only does my suggestion seem to address the issue in @xperjon's reproducer, but when I implement the same suggestion I provided earlier against your submitted PR #974 (patch below), both test cases in your `MtomTest` unit test pass fine.

I haven't dedicated enough time to dig in too deep - but curious why using the CXF HTTP/2 client seems to address the issue; I'll certainly keep an eye on #973 - but I'm left wondering whether your unit test may need some tweaking - ie. assuming it's not tied to HTTP/2, then the patch below should still cause your unit test to fail, no?

```patch
Index: integration-tests/mtom/pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/integration-tests/mtom/pom.xml b/integration-tests/mtom/pom.xml
--- a/integration-tests/mtom/pom.xml	(revision 1e1f04c5b4d1eee2b579c1da496eb2c816c6c397)
+++ b/integration-tests/mtom/pom.xml	(date 1692492962808)
@@ -21,7 +21,10 @@
             <groupId>io.quarkiverse.cxf</groupId>
             <artifactId>quarkus-cxf-rt-features-logging</artifactId>
         </dependency>
-
+        <dependency>
+            <groupId>io.quarkiverse.cxf</groupId>
+            <artifactId>quarkus-cxf-rt-transports-http-hc5</artifactId>
+        </dependency>
         <dependency>
             <groupId>io.rest-assured</groupId>
             <artifactId>rest-assured</artifactId>
Index: integration-tests/mtom/src/test/java/io/quarkiverse/cxf/it/ws/mtom/server/MtomTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/integration-tests/mtom/src/test/java/io/quarkiverse/cxf/it/ws/mtom/server/MtomTest.java b/integration-tests/mtom/src/test/java/io/quarkiverse/cxf/it/ws/mtom/server/MtomTest.java
--- a/integration-tests/mtom/src/test/java/io/quarkiverse/cxf/it/ws/mtom/server/MtomTest.java	(revision 1e1f04c5b4d1eee2b579c1da496eb2c816c6c397)
+++ b/integration-tests/mtom/src/test/java/io/quarkiverse/cxf/it/ws/mtom/server/MtomTest.java	(date 1692493025669)
@@ -9,6 +9,9 @@
 import jakarta.activation.DataHandler;
 import jakarta.xml.ws.Service;
 
+import org.apache.cxf.Bus;
+import org.apache.cxf.frontend.ClientProxy;
+import org.apache.cxf.transport.http.asyncclient.hc5.AsyncHTTPConduit;
 import org.assertj.core.api.Assertions;
 import org.jboss.logging.Logger;
 import org.junit.jupiter.api.Test;
@@ -78,6 +81,10 @@
         final Service service = jakarta.xml.ws.Service.create(serviceUrl, qName);
         final MtomService proxy = service.getPort(MtomService.class);
 
+        Bus bus = ClientProxy.getClient(proxy).getBus();
+        bus.setProperty(AsyncHTTPConduit.USE_ASYNC, Boolean.TRUE);
+        bus.setProperty(AsyncHTTPConduit.ENABLE_HTTP2, Boolean.TRUE);
+        
         DataHandler dh = new DataHandler(new RandomBytesDataSource(size));
         DHResponse response = proxy.echoDataHandler(new DHRequest(dh));
         Assertions.assertThat(response).isNotNull();
```

**Reply by famod** [reply]
Funny enough, @mickroll and I have been seeing some weird http/2 issues after updating to quarkus-cxf 2/CXF 4/Quarkus 3 and your suggestion @shumonsharif to try hc5 async does fix our issues too!

To give more detail: _Without_ hc5 async we were seeing isues like missing request bodies and other low level stream issues when running our app as a client against one of our wiremock mock backends.
There seems to be some issue with transfer-encoding chunked which doesn't even exist anymore in http/2 but we were seeing that after the "auto" upgrade from http/1.1 to http/2 our app was still transmitting that http/1-only header. The issues was only occuring for requests beyond the (default) chunking size (which is 4k or 8k, I'd need to check).
We were able to work around that by forcing http/1.1 when running against http endpoints. Interestingly, when calling the real backends via https, the upgrade to http/2 seems to work fine. Maybe that's because of ALPN, don't know really.

**Reply by famod** [reply]
Turns out that our client-side issue (see my previous comment) is gone after #998 / QCXF 2.3.1 without doing anything, so we probably had hit a bug in `java.net.http.HttpClient`?
I'm referring to https://github.com/quarkiverse/quarkus-cxf/pull/998/files#diff-a1ffd5525feed5ed9a6af2c0c5cdbf0a717f882639c92ea44869e625254634e9R1361-R1366
Does that make sense @ppalaga?

**Reply by ppalaga** [reply]
I am a bit lost which issues you observed on client or server side and which of those clients and servers were actually quarkus-cxf. Anyway, the root cause of the current issue is on quarkus-cxf server side, if the endpoint is plain http and the client is able to speak http/2. Hence switching to a HTTP/1.1 only client which https://github.com/quarkiverse/quarkus-cxf/pull/998/files#diff-a1ffd5525feed5ed9a6af2c0c5cdbf0a717f882639c92ea44869e625254634e9R1361-R1366 does may workaround the issue. I would not say this proves that `java.net.http.HttpClient` was broken.

**Reply by famod** [reply]
Quarkus server side can be ruled out in my case, it's quarkus-cxf client -> wiremock server (which is based on jetty) with upgrade from http 1.1 to 2. Also, it makes no difference whether I use Quarkus 3.3.2 or 3.4.0 (which contains that vert.x fix), unsurprisingly.
> I would not say this proves that `java.net.http.HttpClient` was broken.

Yeah, could also be the conduit's fault.

**Reply by famod** [reply]
FWIW, problem persists with CXF 4.0.3 (using the CxfDefault conduit factory).

**Reply by ppalaga** [reply]
> problem persists with CXF 4.0.3 (using the CxfDefault conduit factory).

Are you calling `close()` on the clients?

**Reply by famod** [reply]
@ppalaga no and TBH I don't see how that is related

**Reply by ppalaga** [reply]
> TBH I don't see how that is related

No, it's not, sorry. I thought you might suffer from https://github.com/quarkiverse/quarkus-cxf/issues/992 but I see that you apparently don't (after reading your posts properly).

**Comment by shumonsharif** [comment]
Hmm, I just tried a simple test, I commented out the following in my patch.

```java
+        bus.setProperty(AsyncHTTPConduit.ENABLE_HTTP2, Boolean.TRUE);
```

And your unit test cases still pass; so looks like it is indeed unrelated to HTTP/2, and likely the async client is doing something differently than the sync one.  Look forward to what you come up with for this issue!

**Reply by famod** [reply]
You need to check what's happening on the wire - there could still be an upgrade to http/2 happening transparently.

**Reply by shumonsharif** [reply]
Very good point; never seem to have enough time these days to truly dig in  :-)  I think there's a good chance that is what's happening ...

**Comment by ppalaga** [comment]
That's very interesting, thanks for investigating @shumonsharif!

**Reply by xperjon** [reply]
Thanks a lot @ppalaga @shumonsharif! I will follow the created issues.

**Comment by dufoli** [comment]
Hello, I remember that a cxf vert request réponse I have written have a buffer management with size. Maybe it need some ajustement

**Comment by ppalaga** [comment, marked as answer]
The root cause is fixed in Vert.x 4.4.5 - see https://github.com/quarkiverse/quarkus-cxf/issues/973
